File: 0021-merge-two-sorted-lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2){
        ListNode aH = null;
        ListNode aT = null;
        ListNode i = list1;
        ListNode j = list2;
        if(list1==null && list2==null){
            return null;
        }
        if(list1!=null && list2==null){
            return list1;
        }
        if(list1==null && list2!=null){
            return list2;
        }
        while(i!=null && j!=null){
            if(i.val<=j.val){
                if(aH==null){
                    aH = i;
                    aT = i;
                } else{
                    aT.next = i;
                    aT = i; 
                }
                i=i.next;
            } else{
                if(i.val>j.val){
                if(aH==null){
                    aH = j;
                    aT = j;
                } else{
                    aT.next = j;
                    aT= j; 
                }
                j=j.next;

            }
        }
        }
        
        if(j!=null){
            aT.next = j;
        }
        if(i!=null){
            aT.next = i;
        }
        return aH;
    }
};

File: 0021-merge-two-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* aH = NULL;
        ListNode* aT = NULL;
        ListNode* i = list1;
        ListNode* j = list2;
        if(list1==NULL && list2==NULL){
            return NULL;
        }
        if(list1!=NULL && list2==NULL){
            return list1;
        }
        if(list1==NULL && list2!=NULL){
            return list2;
        }

        while(i!=NULL && j!=NULL){
            if(i->val<=j->val){
                if(aH==NULL){
                    aH = i;
                    aT = i;
                } else{
                    aT->next = i;
                    aT = i;
                }
                i=i->next;
            } else{
                if(i->val>j->val){
                    if(aH==NULL){
                        aH = j;
                        aT = j;
                    } else{
                        aT->next = j;
                        aT = j;
                    }
                    j=j->next;
                }
            }
        }
        if(i!=NULL){
            aT->next=i;
        }
        if(j!=NULL){
            aT->next=j;
        }
        return aH;
    }
};
